src/support.js

```diff
@@ -0,0 +1,124 @@
+import ReactDOM from "react-dom";
+import copy from "copy-to-clipboard";
+import { detectOS } from "detect-browser";
+
+import styles from "./assets/stylesheets/support.scss";
+import configs from "./utils/configs";
+
+const SHORTHAND_INITIALIZER = "var foo = 'bar'; var baz = { foo };";
+const SPREAD_SYNTAX = "var foo = {}; var baz = { ...foo };";
+const CATCH_SYNTAX = "try { foo(); } catch {}";
+
+function syntaxSupported(syntax) {
+  try {
+    eval(syntax);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+function getPlatformSupport() {
+  return [
+    { name: "Web Assembly", supported: !!window.WebAssembly },
+    { name: "Media Devices", supported: !!navigator.mediaDevices },
+    { name: "Shorthand initializer syntax", supported: syntaxSupported(SHORTHAND_INITIALIZER) },
+    { name: "Spread syntax", supported: syntaxSupported(SPREAD_SYNTAX) },
+    { name: "Optional catch syntax", supported: syntaxSupported(CATCH_SYNTAX) }
+  ];
+}
+
+function isInAppBrowser() {
+  // Some apps like Twitter, Discord and Facebook on Android and iOS open links in
+  // their own embedded preview browsers.
+  //
+  // On iOS this WebView does not have a mediaDevices API at all, but in Android apps
+  // like Facebook, the browser pretends to have a mediaDevices, but never actually
+  // prompts the user for device access. So, we show a dialog that tells users to open
+  // the room in an actual browser like Safari, Chrome or Firefox.
+  //
+  // Facebook Mobile Browser on Android has a userAgent like this:
+  // Mozilla/5.0 (Linux; Android 9; SM-G950U1 Build/PPR1.180610.011; wv) AppleWebKit/537.36 (KHTML, like Gecko)
+  // Version/4.0 Chrome/80.0.3987.149 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/262.0.0.34.117;]
+  const detectedOS = detectOS(navigator.userAgent);
+  return (detectedOS === "iOS" && !navigator.mediaDevices) || /\bfb_iab\b/i.test(navigator.userAgent);
+}
+
+export function platformUnsupported() {
+  return getPlatformSupport().some(s => !s.supported) || isInAppBrowser();
+}
+
+class Support extends React.Component {
+  state = { showDetails: false, copyText: "copy" };
+  toggleDetails = e => {
+    e.preventDefault();
+    this.setState(state => {
+      state.showDetails = !state.showDetails;
+      return state;
+    });
+  };
+  onCopyClicked = e => {
+    e.preventDefault();
+    copy(document.location);
+    this.setState({ copyText: "copied!" });
+  };
+  render() {
+    const platformSupport = getPlatformSupport();
+    const allSupported = platformSupport.every(s => s.supported);
+    const inAppBrowser = isInAppBrowser();
+
+    if (allSupported && !inAppBrowser) return null;
+
+    const detectedOS = detectOS(navigator.userAgent);
+
+    return (
+      <div className={styles.supportMain}>
+        <div className={styles.supportContent}>
+          <div>
+            <img className={styles.logo} src={configs.image("logo")} />
+          </div>
+          <p>
+            Your browser is missing required features.<br />
```

Comment:
Can `messages` from `utils/i18n.js` be used here, or does that run the risk of having the page break? (I suspect it breaks if its dependency, `utils/configs.js`, uses an unsupported language feature)

Reply:
Good point, I'll see if it's possible to use i18n here.